ydb/public/sdk/cpp/client/ydb_federated_topic/impl/federated_write_session.cpp

```diff
@@ -129,29 +129,51 @@ void TFederatedWriteSessionImpl::OpenSubSessionImpl(std::shared_ptr<TDbInfo> db)
     CurrentDatabase = db;
 }
 
-std::shared_ptr<TDbInfo> TFederatedWriteSessionImpl::SelectDatabaseImpl() {
-    std::vector<std::shared_ptr<TDbInfo>> availableDatabases;
+std::pair<std::shared_ptr<TDbInfo>, EStatus> TFederatedWriteSessionImpl::SelectDatabaseImpl() {
+    auto isPreferred = [&preferred = Settings.PreferredDatabase_](const TDbInfo& db) {
+        return AsciiEqualsIgnoreCase(*preferred, db.name()) || AsciiEqualsIgnoreCase(*preferred, db.id());
+    };
+
+    auto const AVAILABLE = TDbInfo::Status::DatabaseInfo_Status_AVAILABLE;
+
+    if (Settings.PreferredDatabase_ && !Settings.AllowFallback_) {
+        // Search for a particular database.
+        for (const auto& db : FederationState->DbInfos) {
+            if (isPreferred(*db)) {
+                return db->status() == AVAILABLE
+                    ? std::pair{db, EStatus::SUCCESS}
+                    : std::pair{nullptr, EStatus::UNAVAILABLE};
+            }
+        }
+        return {nullptr, EStatus::NOT_FOUND};
+    }
+
     ui64 totalWeight = 0;
+    std::vector<std::shared_ptr<TDbInfo>> availableDatabases;
+    std::shared_ptr<TDbInfo> sameDatacenterDb;
 
     for (const auto& db : FederationState->DbInfos) {
-        if (db->status() != TDbInfo::Status::DatabaseInfo_Status_AVAILABLE) {
+        if (db->status() != AVAILABLE) {
             continue;
         }
-
-        if (Settings.PreferredDatabase_ && (AsciiEqualsIgnoreCase(db->name(), *Settings.PreferredDatabase_) ||
-                                            AsciiEqualsIgnoreCase(db->id(), *Settings.PreferredDatabase_))) {
-            return db;
-        } else if (AsciiEqualsIgnoreCase(FederationState->SelfLocation, db->location())) {
-            return db;
+        if (Settings.PreferredDatabase_ && isPreferred(*db)) {
+            return {db, EStatus::SUCCESS};
+        }
+        if (AsciiEqualsIgnoreCase(FederationState->SelfLocation, db->location())) {
+            // Don't return it immediately, as we may find preferred database on next iterations.
+            sameDatacenterDb = db;
         } else {
             availableDatabases.push_back(db);
             totalWeight += db->weight();
         }
     }
 
+    if (sameDatacenterDb) {
```

Comment:
This is not right in case of unavailable preferrableDatabase.  We must route client by hash, but we will use sameDatacanterDb always. It will we correct if we will use desiredDb = preferrableDatabase ? preferableDatabase : sameDatacenterDb. Ignoring Available flag. And then check if desiredDb is available.